clh: Improve logging for cloud hypervisor

jcvenegas opened the pull request:
Send virtiofsd logs to syslog in the same way that qemu implementation
does. This requires not to wait for messages from virtiofsd stdout. This
takes the qemu implementation approach. Give the socket fd to virtiofsd.

Use systemd-cat to collect hypervisor output. The `systemd-cat` program
will open a journal fd and call `cat(1)` to redirect all the output to
the fd. This requires an extra binary to read from hypervisor stdout
(that has combined stdin, stderr and serial terminal). But because it is
cat the overhead is minimal and only is started on Kata debug mode.

Fixes: #2439

jcvenegas commented:
Example,deliberately add wrong option value to hypervisor args
```
docker run -ti --rm --runtime kata-runtime fedora bash
docker: Error response from daemon: OCI runtime create failed: failed to launch cloud-hypervisor: 'Failed to connect to API (timeout 10s): error: Get http://localhost/api/v1/vmm.ping: dial unix /run/vc/vm/1cea95b732c21488e0bbe792bd3107ce9fcbd4a0498ecd1d05650d89cf8fbda9/clh-api.sock: connect: no such file or directory reason: ', hypervisor output:
cloud-hypervisor v0.6.0-287-g3eaeba4
The Cloud Hypervisor Authors
Launch a cloud-hypervisor VMM.

USAGE:
    cloud-hypervisor [FLAGS] [OPTIONS]

FLAGS:
    -h, --help       Prints help information
    -v               Sets the level of debugging output
    -V, --version    Prints version information
...

```

Now running correctly, but going to a guest going to panic because of the wrong kernel version:

```
docker run -ti --rm --runtime kata-runtime fedora bash
docker: Error response from daemon: OCI runtime create failed: Failed to check if grpc server is working: context deadline exceeded: unknown.
```

```
journalctl -t kata-hypervisor -f --output-fields=MESSAGE -o cat
...
[    2.180980]  ? printk+0x3e/0x46
[    2.183955]  mount_block_root+0x27e/0x331
[    2.187697]  prepare_namespace+0x15b/0x18a
[    2.190992]  ? rest_init+0xa0/0xa0
[    2.193404]  kernel_init+0x5/0x100
[    2.196081]  ret_from_fork+0x35/0x40
[    2.199523] Kernel Offset: disabled
[    2.202846] Rebooting in 1 seconds..
VMM thread failed VmReboot(DeviceManager(CreateVirtioFs(VhostUserCreateMaster(VhostUserProtocol(SocketConnect(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }))))))

```

Filter cloud-hypervisor logs:
```
journalctl -t kata-hypervisor -f --output-fields=MESSAGE -o cat | grep 'cloud-hypervisor:'
cloud-hypervisor: 11.392470421s: DEBUG:vm-virtio/src/vsock/device.rs:161 -- vsock: epoll_handler::process_tx()
cloud-hypervisor: 11.392545025s: DEBUG:vm-virtio/src/vsock/unix/muxer.rs:201 -- vsock: muxer.send[rxq.len=0]: [3, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 126, 3, 0, 64, 47, 0, 0, 0, 1, 0, 5, 0, 0, 0, 0, 0, 0, 0, 4, 0, 225, 3, 0, 0]
cloud-hypervisor: 11.392719935s: DEBUG:vm-virtio/src/vsock/device.rs:105 -- vsock: raising IRQ
cloud-hypervisor: 11.39280674s: DEBUG:vm-virtio/src/vsock/device.rs:337 -- vsock: backend event
cloud-hypervisor: 11.392832341s: DEBUG:vm-virtio/src/vsock/unix/muxer.rs:285 -- vsock: muxer received kick
cloud-hypervisor: 11.392862943s: DEBUG:vm-virtio/src/vsock/unix/muxer.rs:357 -- vsock: muxer processing event: fd=123, evset=EPOLLIN
cloud-hypervisor: 11.392905445s: DEBUG:vm-virtio/src/vsock/device.rs:161 -- vsock: epoll_handler::process_tx()
cloud-hypervisor: 11.392946648s: DEBUG:vm-virtio/src/vsock/device.rs:119 -- vsock: epoll_handler::process_rx()
cloud-hypervisor: 11.393017052s: DEBUG:vm-virtio/src/vsock/unix/muxer.rs:178 -- vsock muxer: RX pkt: [2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 126, 3, 0, 64, 0, 4, 0, 0, 17, 0, 0, 0, 1, 0, 5, 0, 0, 0, 0, 0, 0, 0, 1, 0, 136, 1, 0, 0]
cloud-hypervisor: 11.393156059s: DEBUG:vm-virtio/src/vsock/device.rs:105 -- vsock: raising IRQ
```

jcvenegas commented:
/test

jcvenegas commented:
/test

jcvenegas commented:
/test-clh

likebreath commented on virtcontainers/clh.go:
```diff
@@ -877,45 +856,61 @@ func (clh *cloudHypervisor) getAvailableVersion() error {
+		cmdHypervisor.Env = append(cmdHypervisor.Env, "RUST_BACKTRACE=full")
+	}
+
+	hypervisoOutput, err := cmdHypervisor.StdoutPipe()
```
Nit: `hypervisoOutput` -> `hypervisorOutput`

likebreath commented on virtcontainers/clh.go:
```diff
@@ -877,45 +856,61 @@ func (clh *cloudHypervisor) getAvailableVersion() error {
+		cmdHypervisor.Env = append(cmdHypervisor.Env, "RUST_BACKTRACE=full")
+	}
+
+	hypervisoOutput, err := cmdHypervisor.StdoutPipe()
+	if err != nil {
+		return "", -1, err
 	}
+	cmdHypervisor.Stderr = cmdHypervisor.Stdout
```
Also a question: is it safe to open a pipe with `stdoutPipe()` but not read from it (when debug is off)?

It looks like we are ignoring the stdout/stderr of CLH anyway when debug is off. Will it be the same if we open the pipe and redirect `cmdHypervisor.Stdour/Stderr` to it only when debug is on ( and do noting to the stderr/out when debug is off)?

jcvenegas replied:
It is a good question, I think the system has a buffer where it dumps data, in the case when is ignore I imagine new prints to stdout, stderr will get a try again error, not sure how rust handle this. @sboeuf any suggestion here?

jodh-intel commented on virtcontainers/virtiofsd_test.go:
```diff
@@ -64,6 +65,7 @@ func TestVirtiofsdStart(t *testing.T) {
 				},
 			}
 			var ctx context.Context
+			err = os.MkdirAll(filepath.Dir(v.socketPath), testDirMode)
```
As shown on the PR, passing invalid args would be a good test that this is working. Could you find a way to create a unit test for that to guarantee debugging always works (distressing if it doesn't - when you need it, you need it ;)

jcvenegas replied:
Actually, related with this specific case, I needed to create the directory before if not the start function will fail because the runtime will fail open the socket for virtiofsd, in the use case the directory already exists, I wonder but I wonder if the directory does not exist the function should create before open the socket, what do you think?

jodh-intel replied:
Sorry - I'm not commenting on this `mkdir` specifically, I'm just saying that if it's possible, it would be good to add a test to ensure debug output is actually captured from clh/virtiofs. Using some invalid params is one way, but there may be better ones ;)

sboeuf commented on virtcontainers/clh.go:
```diff
@@ -877,45 +856,61 @@ func (clh *cloudHypervisor) getAvailableVersion() error {
-		}
-		args = append(args, cscLogFile)
-		args = append(args, logfile)
+		args = append(args, "-vvv")
```
There's no such thing as `--verbose-max` for now (maybe that's something we can add in the future). The way it works, each `v` gives you more logs, and over 3 `v`s, you get the `Trace` type of logs.
This is not expected to change, but we didn't reach 1.0.0 yet, so for the time being, we'll have to adapt Kata is anything changes in Cloud-Hypervisor (which of course we're trying to avoid as much as possible :))

jcvenegas commented:
/test-clh

jcvenegas commented on virtcontainers/utils/utils.go:
```diff
@@ -275,3 +276,23 @@ const (
 )
+
+// Binary to use to log program output
+const LoggerBinryName = "systemd-cat"
```
It will print the output in the format of cloud hypervisor stdout, I am afraid is not structured logged. 
```
journalctl -t kata-hypervisor -f
Apr 28 21:11:15 jcvenega kata-hypervisor[102976]: cloud-hypervisor: 10.342620707s: DEBUG:vm-virtio/src/vsock/device.rs:161 -- vsock: epoll_handler::process_tx()
Apr 28 21:11:15 jcvenega kata-hypervisor[102976]: cloud-hypervisor: 10.342638908s: DEBUG:vm-virtio/src/vsock/device.rs:119 -- vsock: epoll_handler::process_rx()
Apr 28 21:11:15 jcvenega kata-hypervisor[102976]: cloud-hypervisor: 10.34267121s: DEBUG:vm-virtio/src/vsock/unix/muxer.rs:178 -- vsock muxer: RX pkt: [2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 51, 4, 0, 64, 0, 4, 0, 0, 17, 0, 0, 0, 1, 0, 5, 0, 0, 0, 0, 0, 0, 0, 1, 0, 231, 1, 0, 0]
Apr 28 21:11:15 jcvenega kata-hypervisor[102976]: cloud-hypervisor: 10.342782016s: DEBUG:vm-virtio/src/vsock/device.rs:105 -- vsock: raising IRQ
```

jcvenegas replied:
@jodh-intel @sboeuf I wonder if it worth I take a look to structured log from kata-agent in rust( I think now it supports it, rigth?) and I try to use it in cloud-hypervisor

jodh-intel replied:
Yes, the new rust agent already uses structured logging. If clh wants to, they could just grab:

- https://github.com/kata-containers/kata-containers/tree/master/src/agent/logging
  (see also https://github.com/kata-containers/kata-containers/issues/46).

amshinde commented on virtcontainers/virtiofsd.go:
```diff
@@ -125,11 +150,23 @@ func (v *virtiofsd) args() ([]string, error) {
+		// shared directory tree
 		"-o", "source=" + v.sourcePath,
-		"-o", "vhost_user_socket=" + v.socketPath,
+		// fd number of vhost-user socket
```
why vhost-user, isnt FdSocketNumber  just a unix socket?

jcvenegas replied:
Yes it is both, it is unix socket for vhost user connection, I wanted to keep the same as virtiofsd help `virtiofsd --help`

amshinde commented on virtcontainers/virtiofsd.go:
```diff
@@ -125,11 +150,23 @@ func (v *virtiofsd) args() ([]string, error) {
 		"-o", "no_posix_lock",
+		// shared directory tree
 		"-o", "source=" + v.sourcePath,
-		"-o", "vhost_user_socket=" + v.socketPath,
```
Not looked into what `vhost_user_socket` was used for, but why is this removed here ?

jcvenegas replied:
we have two options to provide the socket path and virtiofsd will create the socket or provide the file descriptor. Because no more stdout check is done by the runtime we require that the hypervisor has access to the socket when it starts. Using the `FD` let us do that without keeping checking until the socket appears.

jcvenegas commented on virtcontainers/clh.go:
```diff
@@ -877,45 +856,61 @@ func (clh *cloudHypervisor) getAvailableVersion() error {
-			cmd.Process.Kill()
+	err = utils.StartCmd(cmdHypervisor)
+	if err != nil {
+		return "Failed to start hypervisor", 0, errors.Errorf("Failed to start hypervisor: %s", err)
```
Thanks, agree with `-1`, I removed the message from `output`

jcvenegas commented:
/test-clh

amshinde commented on virtcontainers/clh.go:
```diff
@@ -372,17 +364,12 @@ func (clh *cloudHypervisor) startSandbox(timeout int) error {
-	}
-	clh.state.PID = pid
-
-	if err := clh.waitVMM(clhTimeout); err != nil {
```
`waitVMM` is no longer required here?

jcvenegas replied:
It was moved to [launchClh](https://github.com/kata-containers/runtime/pull/2648/files#diff-f1c6a53125ae38d7dbb5756ebea5771fR898) if fails to wait for the hypervisor, it read the output of the command and return it as part of the error output.

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh-docker

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-ubuntu

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-ubuntu

likebreath commented:
Seems to be a timeout for a k8s test (because of lots of logging)?
```
09:36:32 not ok 1 Check PID namespaces
09:36:32 # (in test file k8s-pid-ns.bats, line 25)
09:36:32 #   `kubectl wait --for=condition=Ready pod "$pod_name"' failed
09:36:32 # INFO: k8s configured to use runtimeclass
09:36:32 # pod/busybox created
09:36:32 # error: timed out waiting for the condition on pods/busybox
09:36:32 # pod "busybox" deleted
09:36:32 Failed at 90: bats "${K8S_TEST_ENTRY}"
```

jcvenegas commented:
@likebreath thanks for spot that, yeah that is what I am looking I printed both kata-runtime and hypervisor logs to see if just timeout something  else

likebreath commented:
All green. Ready to be merged? 😄

devimc commented:
ehmm.. clh CI is not happy :sweat_smile:

jcvenegas commented:
I have tried to re-run to confirm was OK, I feel tha PR could bring some inestability to the CI, I need more time to confirm if the CI itself is not stable or this PR could make it worst.

jcvenegas commented:
/test-ubuntu

jcvenegas commented:
/test-clh

jcvenegas commented:
/test-ubuntu

jcvenegas commented:
/test-clh

jcvenegas commented:
Ready to be merged found that what causes the instability is a missing kernel patch. I added a depends-on to test the PR and seems to be OK. 

Update: 
- I reduced the logging of the guest if needed we can still add it manually.
- Changed to get hypervisor stdout only on debug mode also to get initial setup, to avoid leave the stdout open but not be readed in that case golang will redirect to `/dev/null`

grahamwhaley commented:
well, we have enough acks and @jcvenegas is happy with it... I'm going to press the big green button.... I think all feedback has been addressed.....
